opm/common/util/numeric/cmp.hpp

```diff
@@ -29,38 +31,101 @@ namespace Opm {
     /// approximate comparison of double values based on absolute
     /// and relative difference. There are three functions:
     ///
-    ///   double_equal() : Compare two double values.
+    ///   scalar_equal<T>() : Compare two <T> values.
     ///
-    ///   double_ptr_equal(): This compares all the element in the
-    ///      two double * pointers.
+    ///   ptr_equal<T>(): This compares all the element in the
+    ///      two T * pointers.
     ///
-    ///   double_vector_equal(): This compares all the elements in
-    ///      two std::vector<double> instances.
+    ///   vector_equal<T>(): This compares all the elements in
+    ///      two std::vector<T> instances.
     ///
-    /// For both double_vector_equal() and double_ptr_equal() the
-    /// actual comparison is based on the double_equal()
+    /// For both vector_equal<T>() and ptr_equal<T>() the
+    /// actual comparison is based on the scalar_equal<T>()
     /// function. All functions exist as two overloads, one which
     /// takes explicit input values for the absolute and relative
-    /// epsilon, and one which uses default values.
+    /// epsilon, and one which uses default values.  
+    ///
+    /// The comparison functions are implemented as templates, with
+    /// the following caveats:
+    ///
+    ///  1. The static_assert() in scalar_equal<T> ensures that only
+    ///     floating point types can be used.
+    ///
+    ///  2. The default epsilon values are of type double -
+    ///     irrespective of the type of data being compared.
     ///
     /// For more details of floating point comparison please consult
     /// this reference:
     ///
     ///    https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition/
+
     namespace cmp {
 
         const double default_abs_epsilon = 1e-8;
         const double default_rel_epsilon = 1e-5;
 
+        template<typename T>
+        bool scalar_equal(T value1, T value2, T abs_eps , T rel_eps) {
+            static_assert(std::is_floating_point<T>::value, "Function scalar_equal() A can only be instantiated with floating point types");
+
+	    bool equal = true;
+            T diff = fabs(value1 - value2);
```

Comment:
> if you don't use a std:: prefix, the plain C version is used

Yes I believe that is true. In (very) theoretical rules-lawyer mode, not using the prefix is not correct C++ anyway, but just happens to work since somewhere upstream math.h has been included.